src/HttpAsyncMessageProducer.php

```diff
@@ -37,22 +39,47 @@ public function __construct(
         parent::__construct($messageConverter, $uri, $requestFactory);
     }
 
-    protected function handleRequest(RequestInterface $request, Deferred $deferred): void
+    protected function handleRequest(RequestInterface $request, Deferred $deferred = null): void
     {
         $promise = $this->httpAsyncClient->sendAsyncRequest($request);
 
+        $exception = null;
+
         $promise->then(
-            function (ResponseInterface $response) use ($deferred) {
-                try {
-                    $payload = $this->getPayloadFromResponse($response);
-                    $deferred->resolve($payload);
-                } catch (\Throwable $exception) {
-                    $deferred->reject($exception);
+            function (ResponseInterface $response) use ($deferred, &$exception) {
+                // we accept only status code 2xx
+                if ('2' !== substr((string) $response->getStatusCode(), 0, 1)) {
+                    if ($deferred) {
+                        $deferred->reject($response->getReasonPhrase());
```

Comment:
We should reject with the full error response so that also body and status code can be inspected because this can contain important information to figure out what went wrong